pkg/k8s/apis/cilium.io/v2/client/register.go

```diff
@@ -201,14 +224,189 @@ func createUpdateCRD(clientset apiextensionsclient.Interface,
+) *apiextensionsv1.CustomResourceDefinition {
+	return &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				k8sconstv2.CustomResourceDefinitionSchemaVersionKey: k8sconstv2.CustomResourceDefinitionSchemaVersion,
+			},
+		},
+		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+			Group: k8sconstv2.CustomResourceDefinitionGroup,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Kind:       template.Spec.Names.Kind,
+				Plural:     template.Spec.Names.Plural,
+				ShortNames: template.Spec.Names.ShortNames,
+				Singular:   template.Spec.Names.Singular,
+			},
+			Scope:    template.Spec.Scope,
+			Versions: template.Spec.Versions,
+		},
+	}
+}
+
+func constructV1beta1CRD(
+	name string,
+	template *apiextensionsv1beta1.CustomResourceDefinition,
+) *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				k8sconstv2.CustomResourceDefinitionSchemaVersionKey: k8sconstv2.CustomResourceDefinitionSchemaVersion,
+			},
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group: k8sconstv2.CustomResourceDefinitionGroup,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Kind:       template.Spec.Names.Kind,
+				Plural:     template.Spec.Names.Plural,
+				ShortNames: template.Spec.Names.ShortNames,
+				Singular:   template.Spec.Names.Singular,
+			},
+			Scope:      template.Spec.Scope,
+			Validation: template.Spec.Validation,
+		},
+	}
+}
+
+func needsUpdateV1(clusterCRD *apiextensionsv1.CustomResourceDefinition) bool {
+	if clusterCRD.Spec.Versions[0].Schema == nil {
+		// no validation detected
+		return true
+	}
+	v, ok := clusterCRD.Labels[k8sconstv2.CustomResourceDefinitionSchemaVersionKey]
+	if !ok {
+		// no schema version detected
+		return true
+	}
+
+	clusterVersion, err := versioncheck.Version(v)
+	if err != nil || clusterVersion.LT(comparableCRDSchemaVersion) {
+		// version in cluster is either unparsable or smaller than current version
+		return true
+	}
+
+	return false
+}
+
+func needsUpdateV1beta1(clusterCRD *apiextensionsv1beta1.CustomResourceDefinition) bool {
+	if clusterCRD.Spec.Validation == nil {
+		// no validation detected
+		return true
+	}
+	v, ok := clusterCRD.Labels[k8sconstv2.CustomResourceDefinitionSchemaVersionKey]
+	if !ok {
+		// no schema version detected
+		return true
+	}
+
+	clusterVersion, err := versioncheck.Version(v)
+	if err != nil || clusterVersion.LT(comparableCRDSchemaVersion) {
+		// version in cluster is either unparsable or smaller than current version
+		return true
+	}
+
+	return false
+}
+
+func convertToV1Beta1CRD(crd *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1beta1.CustomResourceDefinition, error) {
+	internalCRD := new(apiextensions.CustomResourceDefinition)
+	v1beta1CRD := new(apiextensionsv1beta1.CustomResourceDefinition)
```

Comment:
can we move all of these functions into a dedicated file? It will be easier to remove the entire file once we drop support for k8s < 1.16